src/forest.rs

```diff
@@ -124,30 +153,30 @@ where
 
     pub fn all_choices<'a>(
         &'a self,
-        node: Node<'i, P>,
-    ) -> impl Iterator<Item = Node<'i, P>> + Clone + 'a
+        node: Node<'i, G>,
+    ) -> impl Iterator<Item = Node<'i, G>> + Clone + 'a
     where
-        P: 'a,
+        G::NodeKind: 'a,
     {
         match self.grammar.node_shape(node.kind) {
-            NodeShape::Choice => self
-                .possible_choices
+            NodeShape::Choice(_) => self
+                .possibilities
                 .get(&node)
                 .into_iter()
                 .flatten()
                 .cloned()
-                .map(move |kind| Node {
-                    kind,
+                .map(move |choice| Node {
+                    kind: self.grammar.node_shape_choice_get(node.kind, choice),
```

Comment:
This bit here `Node { ... }` is repeated in `one_choice` -- refactor?

Reply:
(I agree, but this pertains to #13)